### ISSUE TITLE
Question: where should the actual data buffers be assigned? 

### ISSUE THREAD

**Issue by bigbrett** [lsgunth/lwip_contrib#1]
First off, thanks for writing this. I'm attempting to port this driver over to the NXP LPC54606, and I think this is a great starting point. 

Is there a reason you never actually assign any of your pbuf's data buffers? The dma descriptors data buffers all are pointing to pbuf->payloads, but you never actually assign pbuf->payload when you allocate them? 

Where would you recommend assigning the actual buffers that the various `pbuf->payload`'s point to?

**Comment by lsgunth** [comment]
You're very welcome, I'm glad somebody is getting some use out of it.

I'm not quite sure what you are referring to... 

On the RX side, the pbufs are created with [pbuf_alloc()](https://www.nongnu.org/lwip/2_0_x/group__pbuf.html#gacfcb0a2af918658ba0afe36499c65f47) in init_rx_dma_desc() which allocates PBUF_POOL_BUFSIZE space for the payload.

On the TX side, we use the pbufs that are passed to us, so the payloads here would be set by the upper layers.

Logan

**Comment by bigbrett** [comment]
Ah, I was under the impression I had to manually point the pbuf->payload to a buffer if I was running lwIP in NO_SYS mode. Thanks for clarifying!

**Comment by bigbrett** [comment]
@lsgunth one more quick question: when you [call netif_add in network_init()](https://github.com/lsgunth/lwip_contrib/blob/1358b5d4f3288ab1ac72a9cbe2d5e64698b2bab9/example/stm32f2x7/src/network.c#L136), you pass it a pointer to ethernet_init() for the 'netif_input_fn' argument (see final argument in the prototype below). 

```
struct netif * netif_add(struct netif *netif, const ip4_addr_t *ipaddr, const ip4_addr_t *netmask, const ip4_addr_t *gw, void *state, netif_init_fn init, netif_input_fn input)
```

 However, it doesn't seem like you have defined that function in this project. I know that is typically the name for the default "process input" function for an lwIP interface that is defined in many of their examples. Is it an issue that you tell the netif that your input function is different than the one that you actually call in the main loop (e.g. `netif_loop()` which calls `stif_loop()`)?

I imagine you could just pass the pointer to this function in netif_add? Or is this benign and has no effect? 

Thanks!
Brett

**Comment by lsgunth** [comment]
I think you misread, it's ethernet_input() not ethernet_init()... The former is part of [lwip](https://www.nongnu.org/lwip/2_0_x/group__lwip__nosys.html#ga6a10c58b82c56d02c48b3cfa2c2494ff) and is one of the standard functions to pass to netif_add().

**Comment by bigbrett** [comment]
Sorry, thats what I meant. Aren't you supposed to define that function though? 

From the comments of netif_add: 

  @ ingroup netif
  Add a network interface to the list of lwIP netifs.
 
  @ param netif a pre-allocated netif structure
  @ param ipaddr IP address for the new netif
  @ param netmask network mask for the new netif
  @ param gw default gateway IP address for the new netif
  @ param state opaque data passed to the new netif
  @ param init callback function that initializes the interface
  **@param input callback function that is called to pass
  ingress packets up in the protocol layer stack.\n
  It is recommended to use a function that passes the input directly
  to the stack (netif_input(), NO_SYS=1 mode) or via sending a
  message to TCPIP thread (tcpip_input(), NO_SYS=0 mode).\n
  These functions use netif flags NETIF_FLAG_ETHARP and NETIF_FLAG_ETHERNET
  to decide whether to forward to ethernet_input() or ip_input().
  In other words, the functions only work when the netif
  driver is implemented correctly!\n**
  Most members of struct netif should be be initialized by the 
  netif init function = netif driver (init parameter of this function).\n
  IPv6: Don't forget to call netif_create_ip6_linklocal_address() after
  setting the MAC address in struct netif.hwaddr
  (IPv6 requires a link-local address).


Edit: I realize now I'm conflating that function with ethernetif_input(), which is a different function written by nxp in their drivers.
